### PR TITLE
[Docs] Provide tables listing basic cython types

### PR DESCRIPTION
**Pull request by matusvalo**
Cython does not provide information about the basic types supported. This PR introduce table showing relation between C types, Cython types and pure python types. Moreover, this PR documents usage of fixed width types.

### CODE REVIEW

**Comment by matusvalo** [comment]
I like also the tables in wikipedia: https://en.wikipedia.org/wiki/C_data_types#Fixed-width_integer_types. Not sure whether this is what we want or not.

**Comment by matusvalo** [comment]
Also what I found undocumented is that in pure python prefixing type with `s` is creating explicitly signed type:
`cython.sint` <-> `signed int`
`cython.sshort` <-> `signed short`
...
No sure whether we want to document (maybe in pure python mode docs?) it since `cython.int` is the same as `cython.sint`. The only real usecase is for `signed char`  <-> `cython.schar`.

**Comment by 0dminnimda** [comment]
> I like also the table in wikipedia documenting fixed width types: https://en.wikipedia.org/wiki/C_data_types#Fixed-width_integer_types. Not sure whether this is what we want or not.

I don't think we need to teach people about c types. It's not the responsibility of cython
We may add a link to this wiki page, but no more

**Comment by matusvalo** [comment]
> I don't think we need to teach people about c types

I agree to some extent. I feel that some very specific types like fastest, pointers etc. should not be explained by Cython documentation. But the basic types should be explained by cython because I consider the goal of cython is: " It makes writing C extensions for Python as easy as Python itself."

So I think that some basic types should be explained by cython to avoid user trying to understand C documentation.

Side not:

My personal opinion is even much radical to be honest. I think that cython should have built-in fixed-with types because also other modern languages have them as built-in:
* https://go.dev/tour/basics/11
* https://doc.rust-lang.org/book/ch03-02-data-types.html#integer-types

**Comment by scoder** [comment]
> My personal opinion is even much radical to be honest. I think that cython should have built-in fixed-with types

I agree and created #4981 for it.

That said, I think mentioning the types in the documentation seems helpful. Explaining them is better done in a C reference or Wikipedia. Wikipedia also has the advantage that it provides translations to many (human) languages.

**Comment by matusvalo** [comment]
> I agree and created https://github.com/cython/cython/issues/4981 for it.

@scoder what should we do with this PR then? I propose then to remove fixed-width types and merge only numeric types. Will you sponsor such PR?

**Comment by matusvalo** [comment]
OK have removed fixed width types so now the PR should be ready for review.

File: docs/src/userguide/language_basics.rst [hunk 1]
```diff
@@ -247,8 +247,61 @@ Types
 
 The Cython language uses the normal C syntax for C types, including pointers.  It provides
 all the standard C types, namely ``char``, ``short``, ``int``, ``long``,
-``long long`` as well as their ``unsigned`` versions,
-e.g. ``unsigned int`` (``cython.uint`` in Python code).
+``long long`` as well as their ``/unsigned`` versions,
```
Comment (da-woods) [inline docs/src/userguide/language_basics.rst:250]: Is the `/` in front of `unsigned` doing anything specific.
?

**Review by da-woods — COMMENTED** [review]
Other than the mystery `\`, looks worth merging to me.

File: docs/src/userguide/language_basics.rst (hunk 1 above)
Comment (matusvalo) [reply docs/src/userguide/language_basics.rst:291]: It's a typo. Let me fix that.

**Comment by matusvalo** [comment]
Hmm there is more types present which I missed when I worked on this PR - e.g. `size_t`, `Py_hash_t`, `Py_ssize_t`, ...

They are mainly coming from https://github.com/cython/cython/blob/master/Cython/Compiler/PyrexTypes.py

I will add them to the table...

**Comment by matusvalo** [comment]
Added additional types, some small issues fixed and added link to stdint.pxd. Let me know if there is something else pending...

**Comment by da-woods** [comment]
Possibly `Py_UCS4` and maybe `Py_UNICODE`. I probably wouldn't mention `Py_UNICODE` though given it's deprecated/removed. But `Py_UCS4` is actually useful.

I don't think it's absolutely important we're comprehensive on the first division though - it's easy to come back and add anything we've missed

**Comment by matusvalo** [comment]
Let me know if there is something missing. If not I will merge it soon.

**Pull request merged**